sycl/include/sycl/ext/intel/experimental/pipes.hpp

```diff
@@ -0,0 +1,213 @@
+#ifdef __SYCL_DEVICE_ONLY__
+    static constexpr auto _anchor_id =
+        _GetValue<int32_t, latency_anchor_id, _Params...>::value;
+    static constexpr auto _constraint =
+        _GetValue3<int32_t, type, int32_t, latency_constraint,
+                   _Params...>::value;
+
+    static constexpr int32_t _target_anchor = std::get<0>(_constraint);
+    static constexpr type _control_type = std::get<1>(_constraint);
+    static constexpr int32_t _cycle = std::get<2>(_constraint);
+    int32_t _type;
+    if (_control_type == type::none) {
+      _type = 0;
+    } else if (_control_type == type::exact) {
+      _type = 1;
+    } else if (_control_type == type::max) {
+      _type = 2;
+    } else { // _control_type == type::min
+      _type = 3;
+    }
+
+    __ocl_RPipeTy<_dataT> _RPipe =
+        __spirv_CreatePipeFromPipeStorage_read<_dataT>(&m_Storage);
+    _dataT TempData;
+    __latency_control_bl_read_wrapper(_RPipe, &TempData, _anchor_id,
+                                      _target_anchor, _type, _cycle);
+    return TempData;
+#else
+    assert(!"Pipes are not supported on a host device!");
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  // Writing to pipe is lowered to SPIR-V instruction OpWritePipe via SPIR-V
+  // friendly LLVM IR.
+  template <class... _Params> static void write(const _dataT &_Data) {
+#ifdef __SYCL_DEVICE_ONLY__
+    static constexpr auto _anchor_id =
+        _GetValue<int32_t, latency_anchor_id, _Params...>::value;
+    static constexpr auto _constraint =
+        _GetValue3<int32_t, type, int32_t, latency_constraint,
+                   _Params...>::value;
+
+    static constexpr int32_t _target_anchor = std::get<0>(_constraint);
+    static constexpr type _control_type = std::get<1>(_constraint);
+    static constexpr int32_t _cycle = std::get<2>(_constraint);
+    int32_t _type;
+    if (_control_type == type::none) {
+      _type = 0;
+    } else if (_control_type == type::exact) {
+      _type = 1;
+    } else if (_control_type == type::max) {
+      _type = 2;
+    } else { // _control_type == type::min
+      _type = 3;
+    }
+
+    __ocl_WPipeTy<_dataT> _WPipe =
+        __spirv_CreatePipeFromPipeStorage_write<_dataT>(&m_Storage);
+    __latency_control_bl_write_wrapper(_WPipe, &_Data, _anchor_id,
+                                       _target_anchor, _type, _cycle);
+#else
+    (void)_Data;
+    assert(!"Pipes are not supported on a host device!");
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+private:
+  static constexpr int32_t m_Size = sizeof(_dataT);
+  static constexpr int32_t m_Alignment = alignof(_dataT);
+  static constexpr int32_t m_Capacity = _min_capacity;
+#ifdef __SYCL_DEVICE_ONLY__
+  static constexpr struct ConstantPipeStorage m_Storage = {m_Size, m_Alignment,
+                                                           m_Capacity};
+
+  template <typename _T>
+  static int32_t
+  __latency_control_nb_read_wrapper(__ocl_RPipeTy<_T> Pipe, _T *Data,
```

Comment:
What is the purpose of this function? Is the idea to depending on if the SPIR-V extension is supported call different SPIR-V instructions inside or something else? I also suggest to make it inline.

Reply:
The purpose is to pass latency control parameters (`int32_t AnchorID, int32_t TargetAnchor, int32_t Type, int32_t Cycle`) to FPGA backend without DSE frontend's support.
In non-experimental version in the future, we will have frontend support so that we will be able to directly pass those parameters through pipe builtin functions.

Reply:
This also explains https://github.com/intel/llvm/pull/5066#discussion_r761971340 . At this case, I'd suggest to put `noinline` attribute instead. Also note, that this design is extremely fragile, for example these function arguments will be deleted, when/if remove dead arg pass is enabled.

Reply:
Is the remove-dead-arg pass a SPIR-V pass? Is there any plan for when will we add it?
We expect to get rid of this experimental version in 2022 Q2 release (oneAPI 2022.3). So as long as we don't have remove-add-arg in 2022 Q1 release (oneAPI 2022.2) I think we are good.